### ISSUE TITLE
Migrating data from one SQLite database to another?

### ISSUE THREAD

**Issue by dingronghui** [groue/GRDB.swift#716]
Hi @groue,
I am a Chinese iOS developer (not so good English). Now I’m developing a fitness app with GRDB. It helps me a lot. Thanks for this great library!

Now i am facing with two problems. I hope you could help me and give me some advice.

1st Migrating existing data from old database into the newest database. I have a  database.db in the App main bundle. The database.db will be copied to user’s documents path when first launched. The db contains table1 which has been adding several basic app records. And the user could also insert or update records to table1. However, when the app is updated, the newest database in the updated App main bundle may already create some new tables.  Table1 ’s column may have been changed and app records of table1 may also been changed . How could I migrate user’s old database’s table1 to the newest database by GRDB?

2nd Do you have any plan to develop CloudKit extension for GRDB? My app needs to sync user’s data with iCloud, such as users’ exercise data and body data and so on. There are several tables need to be synced (they are mapped to CloudKit zones). I have synced  only one table of db because it’s not so hard to follow  the tutorial of CloudKit WWDC. But I think I’m not connecting CloudKit with GRDB very well.I have seen the issues https://github.com/groue/GRDB.swift/issues/547 https://github.com/groue/GRDB.swift/issues/385 .  Are you still developing CloudKit extension for GRDB ?  Maybe I can use GRDB with CloudKit in the near future?
Thank you !

**Comment by groue** [comment]
Hello @dingronghui,

> 1st Migrating existing data from old database into the newest database. I have a database.db in the App main bundle.

So you want to *synchronize* the document database with the resource database.

Synchronization is a classic problem.

Sometimes you want to synchronize with a remote server, and sometimes with an embedded resource database.

It's the same problem. It just happens that the data you want to synchronize with is stored in an embedded resource, instead of a remote server.

So here is how I would do it. When your app boots:

1. [Migrate](https://github.com/groue/GRDB.swift/blob/master/README.md#migrations) your document database up to the latest database schema.
    
    ```swift
    func openDatabase() throws -> DatabaseQueue {
        let documentDatabase = try DatabaseQueue(path: pathToDocumentDatabase)
        let migrator: DatabaseMigrator = ...
        try migrator.migrate(documentDatabase)
    ```

2. Synchronize `table1` with the content of the resource database:

    ```swift
        let resourceDatabase = try DatabaseQueue(path: pathToResourceDatabase)
        try documentDatabase.write { documentDB
            try resourceDatabase.read { resourceDB in
                // Synchronize table1
                let resourceRecords = try MyRecord1.fetchAll(resourceDB)
                let documentRecords = try MyRecord1.fetchAll(documentDB)
                // Insert, delete or update document records according to
                // resource records and your app requirements.
                
                // Synchronize other tables
                ...
            }
        }
    ```

3. Done.
    
    ```swift
        return documentDatabase
    }
    ```


> 2nd Do you have any plan to develop CloudKit extension for GRDB?

Nobody I'm aware of is working on this at the present time.

**Comment by dingronghui** [comment]
@groue Thanks for taking the time to answer my question so thoroughly. Your synchronization method is excellent. And it will help me a lot. Thank you so much.

**Comment by groue** [comment]
Great, happy GRDB!
